api4/system.go

```diff
@@ -364,3 +366,89 @@ func collectMobileTelemetry(c *Context, w http.ResponseWriter, r *http.Request)
 	ReturnStatusOK(w)
 	return
 }
+
+func sendAlertNotification(c *Context, w http.ResponseWriter, r *http.Request) {
+
+	if !c.App.SessionHasPermissionToSendAlerts(c.App.Session) {
+		c.SetPermissionError(model.PERMISSION_MANAGE_JOBS)
+		mlog.Error(fmt.Sprintf("UserId: %v does not have permission to send alerts", c.App.Session.UserId))
+		return
+	}
+
+	notificationReq := model.MobileAlertRequestFromJson(r.Body)
+	if notificationReq == nil {
+		return
+	}
+
+	if notificationReq.PostId == nil || *notificationReq.PostId == "" {
+		mlog.Error("Invalid Request, Mandatory fields are missing - 'postId' ")
+		return
+	}
+
+	postInfo, err := c.App.GetSinglePost(*notificationReq.PostId)
+	if err != nil {
+		mlog.Error(fmt.Sprintf("No valid Post data found for the postId: %v", *notificationReq.PostId))
+		return
+	}
+
+	pushNotificationsEnabled := false
+	if *c.App.Config().EmailSettings.SendPushNotifications {
+		pushServer := *c.App.Config().EmailSettings.PushNotificationServer
+		if license := c.App.License(); pushServer == model.MHPNS && (license == nil || !*license.Features.MHPNS) {
+			mlog.Warn("Push notifications are disabled. Go to System Console > Notifications > Mobile Push to enable them.")
+			pushNotificationsEnabled = false
+		} else {
+			pushNotificationsEnabled = true
+		}
+	}
+
+	channelInfo, _ := c.App.GetChannel(postInfo.ChannelId)
+
+	if *notificationReq.NotifyAllActiveUsers {
+
+		limit := 200
+		offset := 0
+
+		for userSessions, err := c.App.GetAllSessionsWithActiveDeviceIds(limit, offset); err == nil && len(userSessions) > 0; {
+			sendNotificationForActiveSessions(c, userSessions, notificationReq, pushNotificationsEnabled,
+				channelInfo, postInfo)
+			offset = offset + limit
+		}
+
+	} else {
+
+		for _, userId := range *notificationReq.UserIds {
+			userSessions, err := c.App.GetSessions(userId)
+			if err != nil {
+				c.Log.Error("Unable to get sessions for userId:"+userId, mlog.Err(err))
+				continue
+			}
+
+			mlog.Info(fmt.Sprintf("Checking the active sessions for userId: %v", userId))
+			sendNotificationForActiveSessions(c, userSessions, notificationReq, pushNotificationsEnabled,
+				channelInfo, postInfo)
+
+		}
+	}
+
+	ReturnStatusOK(w)
+}
+
+func sendNotificationForActiveSessions(c *Context, userSessions []*model.Session, notificationReq *model.MobileAlertRequest, pushNotificationsEnabled bool,
+	channelInfo *model.Channel, postInfo *model.Post) {
+
+	for _, session := range userSessions {
+
+		if session.IsExpired() {
+			continue
+		}
+
+		deviceId := session.DeviceId
+		userInfo, _ := c.App.GetUser(session.UserId)
+
+		if deviceId != "" && *notificationReq.SendPushNotification && pushNotificationsEnabled {
```

Comment:
I think, if we want to make this endpoint generic, we may want to check the user access to the channel and the post.
Also, Please let me know your thoughts on making the postInfo and Channel Info optional?

Reply:
We can add a validation to check whether the postId belongs to a public channel or not, and throw an error if its related to private channel. 

If we make the postInfo optional, there will be limited message that can be sent in the push notification and user navigation on clicking the notification will needs to be implemented separately as the current push notification structure mandates a link to one of the posts in the channel.